pkg/spire/verify.go

```diff
@@ -74,53 +73,88 @@ func (sc *SpireControllerApiClient) VerifyTaskRunResults(rs []v1beta1.TaskRunRes
 	}
 
 	for key, _ := range resultMap {
-		if strings.HasSuffix(key, ".sig") {
+		if strings.HasSuffix(key, KeySignatureSuffix) {
 			continue
 		}
-		if key == "SVID" {
+		if key == KeySVID {
 			continue
 		}
-		if err := verifyOne(cert.PublicKey, key, resultMap); err != nil {
+		if err := verifyResult(cert.PublicKey, key, resultMap); err != nil {
 			return err
 		}
 	}
 
 	return nil
 }
 
-func hashTaskrunStatus(tr *v1beta1.TaskRun) (string, error) {
-	s, err := json.Marshal(tr.Status)
+// Verify checks if the status has an SVID cert
+// it then verifies the provided signatures against the cert
+func (sc *SpireControllerApiClient) VerifyStatusInternalAnnotation(ctx context.Context, tr *v1beta1.TaskRun, logger *zap.SugaredLogger) error {
+	err := sc.checkClient(ctx)
 	if err != nil {
-		return "", err
+		return err
 	}
-	return fmt.Sprintf("%x", sha256.Sum256(s)), nil
-}
 
-func hashTaskrunStatusInternal(tr *v1beta1.TaskRun) (string, error) {
-	s, err := json.Marshal(tr.Status.TaskRunStatusFields)
+	annotations := tr.Status.Annotations
+
+	// get trust bundle from spire server
+	trust, err := getTrustBundle(sc.workloadAPI, ctx)
 	if err != nil {
-		return "", err
+		return err
 	}
-	return fmt.Sprintf("%x", sha256.Sum256(s)), nil
-}
 
-func CheckStatusAnnotationHash(tr *v1beta1.TaskRun) error {
-	// get stored hash of status
-	hash := tr.Annotations[TaskRunStatusHashAnnotation]
-	// get current hash of status
-	current, err := hashTaskrunStatus(tr)
+	// verify controller SVID
+	svid, ok := annotations[controllerSvidAnnotation]
+	if !ok {
+		return errors.New("No SVID found")
+	}
+	block, _ := pem.Decode([]byte(svid))
+	cert, err := x509.ParseCertificate(block.Bytes)
 	if err != nil {
+		return fmt.Errorf("invalid SVID: %s", err)
+	}
+
+	// verify certificate root of trust
+	if err := verifyCertificateTrust(cert, trust); err != nil {
 		return err
 	}
-	if hash != current {
-		return fmt.Errorf("current status hash and stored annotation hash does not match! Annotation Hash: %s, Current Status Hash: %s", hash, current)
+	logger.Infof("Successfully verified certificate %s against SPIRE", svid)
+
+	if err := verifyAnnotation(cert.PublicKey, annotations); err != nil {
+		return err
 	}
+	logger.Info("Successfully verified signature")
+
+	// check current status hash vs annotation status hash by controller
+	if err := checkStatusInternalAnnotation(tr, annotations); err != nil {
+		return err
+	}
+	logger.Info("Successfully verified status annotation hash matches the current taskrun status")
+
 	return nil
 }
 
-func CheckStatusInternalAnnotationHash(tr *v1beta1.TaskRun) error {
+func (sc *SpireControllerApiClient) SpireVerified(tr *v1beta1.TaskRun) bool {
```

Comment:
I think the naming could lead to some confusion, this could be interpreted as doing the verification. Maybe `CheckSpireVerifiedFlag`?